src/python/pants/option/ranked_value.py

```diff
@@ -1,13 +1,40 @@
 # Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from dataclasses import dataclass
 from enum import Enum
-from typing import Any, Dict, Iterator, List, Optional, Union, cast
+from functools import total_ordering
+from typing import Any, Dict, Iterator, List, Union
+
+
+@total_ordering
+class Rank(Enum):
+    # The ranked value sources. Higher ranks override lower ones.
+    NONE = (0, "NONE")  # The value None.
+    HARDCODED = (1, "HARDCODED")  # The default provided at option registration.
+    CONFIG_DEFAULT = (2, "CONFIG_DEFAULT")  # The value from the DEFAULT section of the config file.
+    CONFIG = (3, "CONFIG")  # The value from the relevant section of the config file.
+    ENVIRONMENT = (4, "ENVIRONMENT")  # The value from the appropriately-named environment variable.
+    FLAG = (5, "FLAG")  # The value from the appropriately-named command-line flag.
+
+    _rank: int
+
+    def __new__(cls, rank: int, display: str) -> "Rank":
```

Comment:
Is your thinking that the complication of providing the display string explicitly is worth it so we can separate the symbolic value from the rank?

Reply:
No real thinking. I was working backwards from IntEnum not being a solution since options parsing expects Enum values to be strings.

Reply:
+1 on removing the display name from the value, if possible. 

You could get the same display name by calling `rank.name`. `rank.value` will give you the int code.

Reply:
> See my explanation to Benjy.

Ah, my GitHub UI was showing the original unedited comment without the full explanation when I wrote this. The explanation sounds reasonable. Thanks!